src/range_proof.rs

```diff
@@ -0,0 +1,555 @@
+        }
+
+        // Compute 2**N-1 for later use
+        let mut two_n_minus_one = Scalar::from(2u8);
+        for _ in 0..log_n {
+            two_n_minus_one = two_n_minus_one * two_n_minus_one;
+        }
+        two_n_minus_one -= Scalar::one();
+
+        // Weighted coefficients for common generators
+        let mut g_base_scalar = Scalar::zero();
+        let mut h_base_scalar = Scalar::zero();
+        let mut gi_base_scalars = vec![Scalar::zero(); max_mn];
+        let mut hi_base_scalars = vec![Scalar::zero(); max_mn];
+
+        // Final multiscalar multiplication data
+        let mut scalars: Vec<Scalar> = vec![];
+        let mut points: Vec<RistrettoPoint> = vec![];
+
+        // Recovered masks
+        let mut masks = vec![];
+
+        // Process each proof and add it to the batch
+        let rng = &mut thread_rng();
+        for (index, proof) in range_proofs.iter().enumerate() {
+            let commitments = statements[index].commitments.clone();
+            let a = proof.get_a()?;
+            let a1 = proof.get_a1()?;
+            let b = proof.get_b()?;
+            let r1 = proof.get_r1();
+            let s1 = proof.get_s1();
+            let d1 = proof.get_d1();
+            let li = proof.get_li()?;
+            let ri = proof.get_ri()?;
+
+            if li.len() != ri.len() {
+                return Err(ProofError::InvalidLength(
+                    "Vector L length not equal to vector R length".to_string(),
+                ));
+            }
+            if 1 << li.len() != commitments.len() * bit_length {
+                return Err(ProofError::InvalidLength(
+                    "Vector L length not adequate".to_string(),
+                ));
+            }
+
+            // Helper values
+            let batch_size = commitments.len();
+            let rounds = li.len();
+
+            // Batch weight (may not be equal to a zero valued scalar)
+            let weight = Scalar::random(rng);
+
+            // Start the transcript
+            let mut transcript = Transcript::new(transcript_label.as_bytes());
+            transcript.domain_separator(b"Bulletproof+", b"Range Proof");
+            transcript.validate_and_append_point(b"H", &h_base.compress())?;
+            transcript.validate_and_append_point(b"G", &h_base.compress())?;
+            transcript.append_u64(b"N", bit_length as u64);
+            transcript.append_u64(b"M", batch_size as u64);
+            for i in 0..statements[index].generators.hi_base().len() {
+                transcript.validate_and_append_point(
+                    b"Hi",
+                    &statements[index].generators.hi_base()[i].compress(),
+                )?;
+            }
+            for i in 0..statements[index].generators.gi_base().len() {
+                transcript.validate_and_append_point(
+                    b"Gi",
+                    &statements[index].generators.gi_base()[i].compress(),
+                )?;
+            }
```

Comment:
Since each proof uses a subset of a common generator set, you could do one of a couple of things to make this more efficient:
1. Leave out these generators, provided they are guaranteed never to change or be under adversarial control. This is a somewhat brittle approach.
2. Precompute domain-separated hashes of each subset (or hardcode them and include a test for their correctness), and use one of these hashes in the transcript instead. This is a safer approach.

Reply:
On further thought and inspection, I think you can safely remove these generator vectors from the transcript, since they can't be externally changed when a statement and proof are presented to the verifier. That being said, it's extremely important that the implementation not change the way these are generated. If this is at all a possibility, you could include a version number in the transcript, and increment it if the generators are ever changed.

I do think the `G` and `H` generators should still be included (even though they are also not specifically presented to the verifier), since this is cheap.